rcl/src/rcl/subscription.c

```diff
@@ -194,6 +214,160 @@ rcl_subscription_get_default_options()
   return default_options;
 }
 
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_subscription_options_copy(
+  const rcl_subscription_options_t * src,
+  rcl_subscription_options_t * dst
+)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(src, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(dst, RCL_RET_INVALID_ARGUMENT);
+  const rcl_allocator_t * allocator = &src->allocator;
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+
+  rcl_ret_t ret = RCL_RET_OK;
+  dst->qos = src->qos;
+  dst->allocator = src->allocator;
+  // copy rmw_subscription_options_t
+  dst->rmw_subscription_options.rmw_specific_subscription_payload =
+    src->rmw_subscription_options.rmw_specific_subscription_payload;
+  dst->rmw_subscription_options.ignore_local_publications =
+    src->rmw_subscription_options.ignore_local_publications;
+  if (src->rmw_subscription_options.filter_expression) {
+    dst->rmw_subscription_options.filter_expression =
+      rcutils_strdup(src->rmw_subscription_options.filter_expression, *allocator);
+    if (!dst->rmw_subscription_options.filter_expression) {
+      ret = RCL_RET_BAD_ALLOC;
+      goto clean;
+    }
+  }
+
+  if (src->rmw_subscription_options.expression_parameters) {
```

Comment:
probably return error if `filter_expression` is not null but `expression_parameters` is null. i believe this case cannot exist.

Reply:
I'll move it inside the above `if` statement.